docs/shared-constraints.md

```diff
@@ -0,0 +1,69 @@
+# Shared Constraints
+
+Custom constraints in `protovalidate` afford a lot of power, but can become
+cumbersome and repetitive when the same kind of custom constraints are needed
+across multiple fields or messages. To this end, `protovalidate` provides a
+mechanism for creating reusable constraints that can be applied on multiple
+fields.
+
+Shared constraints require Protobuf extensions, which are not available in
+proto3. Either proto2 or at least Protobuf 2023 Edition must be used to define
+shared constraints. Shared constraints defined in proto2 or Protobuf 2023
+Edition or later can be imported and utilized in proto3 files.
+
+## Shared Field Constraints
+
+To define a shared field rule, extend one of the standard rules messages. For
+example, to define a new rule for `float` fields, extend
+`buf.validate.FloatRules`, as follows:
+
+```proto
+import "buf/validate/shared/constraints.proto";
+import "buf/validate/validate.proto";
+
+extend buf.validate.FloatRules {
+  float abs_range = 80048952 [(buf.validate.shared.field).cel = {
+    id: "float.abs_range"
+    expression: "this >= -rule && this <= rule"
+    message: "float value is out of range"
+  }];
+}
```

Comment:
This version of the proposal doesn't support recursive constraints, so all you can do for now is custom CEL expressions. It would be nice to support and isn't particularly hard to do, but it required resolving design issues that I think we wanted to punt on for the initial version.

My thought is that we will probably want a new option for this that lets you specify a `FieldConstraint` to be applied as a sort-of template.

There's really a couple design points that need to be hammered out first, though:
- How do you determine _when_ templates should be applied? Currently rules apply when they are considered set, but that would mean that in this example, something doing `(buf.validate.field).string.(valid_path) = false` would have the `string.matches` rule applied to it. So, it seems like there needs to be a CEL expression that determines if the template should be applied (and presumably, it should default to treating the rule field as a boolean, so you don't need to specify it for the `bool` case.)
- There is a bit of redundancy&mdash;you can apply a CEL expression using either: a template with `buf.validate.field.cel`, or using `buf.validate.shared.field.cel`. There are some semantic differences though (custom constraints still don't get `rule` values) and I think we'll just have to live with this fact.

So the design will _probably_ wind up looking something like this:

```proto
extend google.protobuf.FieldOptions {
  // Specifies constraints to recursively apply for a shared constraint rule. These are merged
  // together using Protobuf merging semantics, allowing more specific rules to override less
  // specific ones.
  optional FieldConstraintsTemplate field = 1161;
}

message FieldConstraintsTemplate {
  // Specifies field rules that will be merged when this template rule is applied.
  FieldConstraints constraints;

  // Specifies a CEL expression that determines if this template rule should be applied.
  // If unspecified, defaults to `bool(rule)`, e.g. applies if the field is a true boolean value.
  string if;
};
```

And the resulting usage would be like this:

```proto
extend buf.validate.StringRules {
  bool valid_path = ... [
    (buf.validate.template.field).constraints = {
      string.matches: "^([^/.][^/]?|[^/][^/.]|[^/]{3,})(/([^/.][^/]?|[^/][^/.]|[^/]{3,}))*$"
    }
  ];
}
```

Reply:
Resolving for now, we'll probably want to workshop this design but I'm hopeful it won't take long.

There's a couple of other follow-ups that would be good too (shared _message_ constraints, shared field constraints on arbitrarily-typed fields?)